controller/service.go

```diff
@@ -78,15 +78,40 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 	// none or a malformed one, nuke all controlled state so that we
 	// start converging from a clean slate.
 	for i := range svc.Status.LoadBalancer.Ingress {
-		ip := svc.Status.LoadBalancer.Ingress[i].IP
-		if len(ip) != 0 {
-			lbIPs = append(lbIPs, net.ParseIP(ip))
+		ipStr := svc.Status.LoadBalancer.Ingress[i].IP
+		if len(ipStr) != 0 {
+			ip := net.ParseIP(ipStr)
+			if ip != nil {
+				lbIPs = append(lbIPs, ip)
+			}
+		}
+	}
+	familyPolicy := v1.IPFamilyPolicySingleStack
+	if svc.Spec.IPFamilyPolicy != nil {
+		familyPolicy = *(svc.Spec.IPFamilyPolicy)
+	}
+	// If svc currently has 1 ip and policy PreferDualStack, try assigning ip from the missing family and same pool
+	if len(lbIPs) == 1 && familyPolicy == v1.IPFamilyPolicyPreferDualStack {
+		level.Info(l).Log("event", "tryAssignAdditionalIP", "msg", "familyPolicy is PreferDualStack, trying to assign additional ip")
+		existingFamily := ipfamily.ForAddress(lbIPs[0])
+		additionalFamily := ipfamily.IPv4
+		if existingFamily == ipfamily.IPv4 {
+			additionalFamily = ipfamily.IPv6
+		}
+		currentPool := c.ips.Pool(key)
+		newIP, err := c.ips.AllocateFromPoolForFamily(key, svc, additionalFamily, currentPool, k8salloc.Ports(svc), SharingKey(svc), k8salloc.BackendKey(svc))
+		if err != nil {
+			c.client.Errorf(svc, "AdditionalAssignFailed", "invalid requested loadbalancer IP: %s", err)
+		} else {
+			lbIPs = append(lbIPs, newIP...)
+			level.Info(l).Log("event", "ipAllocated", "ip", newIP, "msg", "Additional IP address assigned by controller")
+			c.client.Infof(svc, "IPAllocated", "Assigned additional IP %q", newIP)
 		}
 	}
 	if len(lbIPs) == 0 {
 		c.clearServiceState(key, svc)
 	} else {
-		lbIPsIPFamily, err := ipfamily.ForAddressesIPs(lbIPs)
+		lbIPsIPFamily, err := ipfamily.ForAddressesIPs(lbIPs, familyPolicy)
```

Comment:
I would expect this to not care about the policy

Reply:
I don't think having just the IPs here is enough. Please check @fedepaol comment at top for a reason why we need this

```
All in all, I think we have a big oversemplification which is: we look at the ipfamily of the service by trying to understand that looking at the clusterips.
```

I understand that you don't want to link the IPFamily with the FamilyPolicy, but as I mentioned above, that would require a really big change, and I don't find any reasons to do so right now?